presto-main/src/main/java/com/facebook/presto/operator/HashBuilderOperator.java

```diff
@@ -577,6 +597,10 @@ private void finishLookupSourceUnspilling()
         spilledLookupSourceHandle.setLookupSource(partition);
 
         state = State.INPUT_UNSPILLED_AND_BUILT;
+        if (isSpillInSmallMemoryPool(operatorContext.getSession())) {
+            unspillInProgress = Optional.empty();
```

Comment:
This line looks like a trivial change, but it is not true.  State `INPUT_UNSPILLING` requires it to be present. That's why it can only be destroyed after state changes to next state. In case there are other dependencies for unspillInProgress,  the code path was controlled by this session property.